File: articles/chemical-biology-molecular-intervention-living-systems/julia/chemical_biology_kernel.jl
# Chemical biology numerical kernel in Julia.

function dose_response(concentration, ec50, hill, bottom, top)
    return bottom + (top - bottom) / (1.0 + (ec50 / concentration)^hill)
end

function occupancy(ligand, kd)
    return ligand / (kd + ligand)
end

function target_engagement(signal_control, signal_treated, signal_max)
    return (signal_control - signal_treated) / (signal_control - signal_max)
end

println("response=", round(dose_response(1.0, 1.5, 1.2, 0.05, 1.0), digits=6))
println("occupancy=", round(occupancy(2.0, 2.0), digits=6))
println("target_engagement=", round(target_engagement(100.0, 55.0, 20.0), digits=6))

File: articles/chemical-biology-molecular-intervention-living-systems/fortran/chemical_biology_kernel.f90
program chemical_biology_kernel
  implicit none

  real(8) :: response, occupancy, engagement

  response = 0.05d0 + (1.0d0 - 0.05d0) / (1.0d0 + (1.5d0 / 1.0d0) ** 1.2d0)
  occupancy = 2.0d0 / (2.0d0 + 2.0d0)
  engagement = (100.0d0 - 55.0d0) / (100.0d0 - 20.0d0)

  print *, "response=", response
  print *, "occupancy=", occupancy
  print *, "target_engagement=", engagement

end program chemical_biology_kernel
